data_structures/src/chain.rs

```diff
@@ -2902,6 +2900,111 @@ impl ReputationEngine {
     pub fn clear_threshold_cache(&self) {
         self.threshold_cache.borrow_mut().clear_threshold_cache()
     }
+
+    /// Get ARS keys ordered by reputation. If tie, order by pkh.
+    pub fn get_rep_ordered_ars_list(&self) -> Vec<PublicKeyHash> {
+        self.ars
+            .active_identities()
+            .cloned()
+            .sorted_by_key(|&pkh| std::cmp::Reverse((self.trs.get(&pkh), pkh)))
+            .collect()
+    }
+}
+
+#[allow(clippy::cast_possible_truncation, clippy::cast_sign_loss)]
+fn magic_line(x: f64, m: f64, k: f64) -> u32 {
+    let res = m * x + k;
+    if res < 0_f64 {
+        0
+    } else {
+        res.round() as u32
+    }
+}
+
+#[allow(clippy::cast_possible_truncation, clippy::cast_precision_loss)]
+fn trapezoid_reputation(
+    sorted_rep: Vec<PublicKeyHash>,
+    trs: &TotalReputationSet<PublicKeyHash, Reputation, Alpha>,
+) -> (u64, Vec<u32>, HashMap<PublicKeyHash, u32>) {
+    let mut total_active_rep = 0;
+    let filtered_identities = sorted_rep
+        .iter()
+        .filter_map(|pkh| {
+            let rep = trs.get(pkh).0;
+            if rep != 0 {
+                total_active_rep += u64::from(rep);
+                Some(rep)
+            } else {
+                None
+            }
+        })
```

Comment:
- This closure could use a comment explaining what's the criteria for filter-mapping.
- This closure can be refactored into a proper function that we can unit test.